src/4D_api.cpp

```diff
@@ -966,6 +966,517 @@ std::string pj_add_type_crs_if_needed(const std::string& str)
+                      const double ymax,
+                      double* out_xmin,
+                      double* out_ymin,
+                      double* out_xmax,
+                      double* out_ymax,
+                      int densify_pts
+) {
+    *out_xmin = HUGE_VAL;
+    *out_ymin = HUGE_VAL;
+    *out_xmax = HUGE_VAL;
+    *out_ymax = HUGE_VAL;
+
+    if (P == nullptr) {
+        proj_log_error(P, _("NULL P object not allowed."));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+    if (densify_pts < 0 || densify_pts > 10000) {
+        proj_log_error(P, _("densify_pts must be between 0-10000."));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+
+    PJ_PROJ_INFO pj_info = proj_pj_info(P);
+    if (pj_info.id == nullptr) {
+        proj_log_error(P, _("NULL transformation not allowed,"));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+    if (strcmp(pj_info.id, "noop") == 0 || direction == PJ_IDENT) {
+        *out_xmin = xmin;
+        *out_xmax = xmax;
+        *out_ymin = ymin;
+        *out_ymax = ymax;
+        return true;
+    }
+
+    bool degree_output = proj_degree_output(P, direction) != 0;
+    bool degree_input = proj_degree_input(P, direction) != 0;
+    if (degree_output && densify_pts < 2) {
+        proj_log_error(P, _("densify_pts must be at least 2 if the output is geograpic."));
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+
+    int side_pts = densify_pts + 1;  // add one because we are densifying
+    const int boundary_len = side_pts * 4;
+    std::vector<double> x_boundary_array;
+    std::vector<double> y_boundary_array;
+    try
+    {
+        x_boundary_array.resize(boundary_len);
+        y_boundary_array.resize(boundary_len);
+    }
+    catch( const std::exception & e ) // memory allocation failure
+    {
+        proj_log_error(P, e.what());
+        proj_errno_set (P, PROJ_ERR_INVALID_OP_ILLEGAL_ARG_VALUE);
+        return false;
+    }
+    double delta_x = 0;
+    double delta_y = 0;
+    bool north_pole_in_bounds = false;
+    bool south_pole_in_bounds = false;
+    bool output_lon_lat_order = false;
+    if (degree_output) {
+        int out_order_lon_lat = target_crs_lon_lat_order(context, P, direction);
+        if (out_order_lon_lat == -1)
+            return false;
+        output_lon_lat_order = out_order_lon_lat != 0;
+        north_pole_in_bounds = contains_north_pole(
+            P,
+            direction,
+            xmin,
+            ymin,
+            xmax,
+            ymax,
+            output_lon_lat_order
+        );
+        south_pole_in_bounds = contains_south_pole(
+            P,
+            direction,
+            xmin,
+            ymin,
+            xmax,
+            ymax,
+            output_lon_lat_order
+        );
+    }
+
+    if (degree_input && xmax < xmin) {
```

Comment:
the source CRS lon/lat order should be tested here

Reply:
Are you thinking that an error should be raised if this condition is true and the axis order is incorrect?

Reply:
> Are you thinking that an error should be raised if this condition is true and the axis order is incorrect?

that would be reasonable